### ISSUE TITLE
Does yolov8 freeze parameter in training works correct? 

### ISSUE THREAD

**Issue by FiksII** [ultralytics/ultralytics#6020]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I am going to use transfer learning with freezing backbones layers. For yolov5, I have good results:

```
python yolov5/train.py --data *****/data.yaml --weights yolov5l.pt --freeze 10 --epochs 300 --batch 8
```
![2023-10-30_11-46-10](https://github.com/ultralytics/ultralytics/assets/42637009/18e9a33d-55e7-467d-885c-b174a098b216)

But, if I use yolov8:

```
from ultralytics import YOLO
model = YOLO('yolov8l.yaml').load('yolov8l.pt')
model.train(data='/****/data.yaml',freeze=10, pretrained=True)
```
It notifies that layers are frozen:
```
Freezing layer 'model.0.conv.weight'
Freezing layer 'model.0.bn.weight'
Freezing layer 'model.0.bn.bias'
Freezing layer 'model.1.conv.weight'
Freezing layer 'model.1.bn.weight'
Freezing layer 'model.1.bn.bias'
Freezing layer 'model.2.cv1.conv.weight'
.........
Freezing layer 'model.9.cv1.bn.weight'
Freezing layer 'model.9.cv1.bn.bias'
Freezing layer 'model.9.cv2.conv.weight'
Freezing layer 'model.9.cv2.bn.weight'
Freezing layer 'model.9.cv2.bn.bias'
Freezing layer 'model.22.dfl.conv.weight'
```

But, the results are similar to training from the scratch.

![2023-10-30_12-06-51](https://github.com/ultralytics/ultralytics/assets/42637009/8dee6342-f3dc-4091-b92b-b75c281238ab)

What am I doing wrong?

**Comment by Kayzwer** [comment]
make sure to set warmup epochs to 0, if not the lr is high in the early phase.

**Comment by FiksII** [comment]
@Kayzwer Thank you for the reply. `warmup_epochs=0` did not fix the problem.

**Comment by glenn-jocher** [comment]
Hi @FiksII,

Thank you for your reply and the suggestion to set `warmup_epochs` to `0`. I'm sorry to hear it did not solve the issue. 

When you perform transfer learning, individual layers are frozen to retain the pre-trained knowledge, so it sounds like you're doing everything correctly. However, as you've observed, the problem seems to be that the model isn't learning from the frozen layers as intended.

When you say that the results are similar to training from scratch, could you please elaborate? Are you referring to the model's final validation performance, training loss progression, or another metric?

The behavior you've described can sometimes happen when the pre-training and fine-tuning datasets have very different statistical properties. If the distributions of the two datasets are not aligned well, the benefits of transfer learning may not be apparent.

Remember that setting the learning rate can have a significant impact on the training of the model as well. If the learning rate is too high, the network may not properly fine-tune the pretrained weights. Generally, a lower learning rate is used in fine-tuning than in training from scratch.

Please continue to experiment with different combinations of hyperparameters and configurations. Your understanding and patience in this matter are highly appreciated.

Let us know if you have further questions.

Ultralytics Team.

**Comment by FiksII** [comment]
@glenn-jocher Thank you for the response. I think I found what is mine concern. It seems yolov8 training metrics (box_loss, cls_loss, dfl_loss) are calculated differently than yolov5.

**_Yolov5 with freeze=10._**
**Train metrics:**
```
box_loss   obj_loss   cls_loss 
 0.07139    0.02851    0.04188
```
**Valid metrics**
```
 Class     Images  Instances          P          R      mAP50   mAP50-95: 
   all       1273       1725      0.619      0.567      0.522      0.241
```

**_Yolov8 with freeze=10._**
**Train metrics:**
```
 box_loss   cls_loss   dfl_loss
  2.052       15.9      2.106
```
**Test metrics:**
```
 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)
    all       1273       1725       0.695      0.698      0.722      0.406
```
**_Yolov8 from scratch**
**Train metrics:**
```
 box_loss   cls_loss   dfl_loss
  3.301        5.419      3.529
```
**Test metrics:**
```
 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)
    all       1273       1725      0.4      0.163     0.0582     0.0177
```

So, precision, recall and mAPs work as I expected, but box_loss, cls_loss, dfl_loss don't.

**Comment by glenn-jocher** [comment]
Hi @FiksII,

Thank you for your detailed analysis and the comprehensive metrics that you have shared.

You're correct in noting that the training metrics for YOLOv8, specifically box_loss, cls_loss, and dfl_loss, may appear differently compared to YOLOv5. YOLOv8 introduces a new approach to object detection, "Double Feature Loss" (DFL), which is likely why you're seeing the additional dfl_loss and altered box_loss and cls_loss values.

The Double Feature Loss, combined with other architectural changes, can affect how the model adjusts its weights throughout training, impacting these losses.

When comparing models, though, it's more important to focus on the evaluation metrics including precision (P), recall (R), mean Average Precision (mAP), etc. These metrics provide a more accurate view of the model's performance on unseen data.

Your test metrics for YOLOv8 with freeze=10 show higher scores compared to the model trained from scratch, which confirms it's learning from the frozen layers.

So, while the training losses in YOLOv8 show a substantial magnitude difference from YOLOv5, it doesn't necessarily mean that something is wrong with the training method, but rather that the loss computation is different due to additional factors (like DFL) in YOLOv8.

Glenn Jocher, Ultralytics Team

**Comment by tahanaveed97** [comment]
I was reviewing a thread and have a question to ask. I've trained a YOLOv8 model using pretrained weights, and also trained it from scratch without freezing any layers. After completing my first version of training, I obtained a new dataset that includes both new images and the images I used in the first training iteration. Now, I wanted to improve my first version training weights.

For this, I used the weights from the first version and set freeze=10. This approach yielded better accuracy compared to fine-tuning on the first version of training weights without freezing. However, training from pretrained weights did not produce as good results.

The new data comes from the same environment. Could you suggest the best training approach for my situation?

Thanks!

**Comment by glenn-jocher** [comment]
@tahanaveed97 hi there,

Great to hear about your progress with YOLOv8! Given that your new dataset includes the original images plus additional ones, and you've seen improved accuracy with freezing the first 10 layers, I'd suggest continuing with this approach. It seems that the frozen layers are providing a stable feature extraction base, while the unfrozen layers adapt to the new data.

Since the new data is from the same environment, the pretrained features should still be relevant. Keep fine-tuning with layer freezing as it's giving you better results. Monitor your validation metrics closely to ensure continued improvement.

Happy training!

Glenn Jocher, Ultralytics Team
